lib/src/modules/task.cc

```diff
@@ -0,0 +1,320 @@
+    processOutputAndUpdateMetadata(response);
+}
+
+void Task::callNonBlockingAction(
+    const ActionRequest& request,
+    const TaskCommand &command,
+    const std::map<std::string, std::string> &environment,
+    const std::string &input,
+    ActionResponse &response
+) {
+    const fs::path &results_dir = request.resultsDir();
+    lth_jc::JsonContainer wrapper_input;
+
+    wrapper_input.set<std::string>("executable", command.executable);
+    wrapper_input.set<std::vector<std::string>>("arguments", command.arguments);
+    wrapper_input.set<std::string>("input", input);
+    wrapper_input.set<std::string>("stdout", (results_dir / "stdout").string());
+    wrapper_input.set<std::string>("stderr", (results_dir / "stderr").string());
+    wrapper_input.set<std::string>("exitcode", (results_dir / "exitcode").string());
+
+    auto exec = lth_exec::execute(
+        wrapper_executable_,
+        {},
+        wrapper_input.toString(),
+        environment,
+        [results_dir](size_t pid) {
+            auto pid_file = (results_dir / "pid").string();
+            lth_file::atomic_write_to_file(std::to_string(pid) + "\n", pid_file);
+        },  // pid callback
+        0,  // timeout
+        leatherman::util::option_set<lth_exec::execution_options> {
+            lth_exec::execution_options::thread_safe,
+            lth_exec::execution_options::create_detached_process,
+            lth_exec::execution_options::merge_environment,
+            lth_exec::execution_options::inherit_locale });  // options
+
+    if (exec.exit_code == TASK_WRAPPER_FILE_ERROR_EC) {
+        // This is unexpected. The output of the task will not be
+        // available for future transaction status requests; we cannot
+        // provide a reliable ActionResponse.
+        std::string empty_label { lth_loc::translate("(empty)") };
+        LOG_WARNING("The task wrapper process failed to write output on file for the {1}; "
+                    "stdout: {2}; stderr: {3}",
+                    request.prettyLabel(),
+                    (exec.output.empty() ? empty_label : exec.output),
+                    (exec.error.empty() ? empty_label : exec.error));
+        throw Module::ProcessingError {
+            lth_loc::translate("failed to write output on file") };
+    }
+
+    // Stdout / stderr output should be on file; read it
+    response.output = storage_.getOutput(request.transactionId(), exec.exit_code);
+    processOutputAndUpdateMetadata(response);
+}
+
+ActionResponse Task::callAction(const ActionRequest& request)
+{
+    auto task_execution_params = request.params();
+    auto task_input_format = task_execution_params.includes("inputFormat") ?
+        task_execution_params.get<std::string>("inputFormat") : std::string{""};
+    std::map<std::string, std::string> task_environment;
+    std::string task_input;
+
+    bool has_input = false;
+    if (task_input_format.empty() || task_input_format == "stdin:json") {
```

Comment:
This doesn't match the current spec which splits format and method. There has been some discussion about simplifying this so it's probably ok for now